### PR TITLE
lesson_5

### PR DESCRIPTION
**Pull request by Reallocman**

### CODE REVIEW

**Review by dimonl — APPROVED** [review]
Need to format all your file. It's hard to read. Try to use 'go fmt ./...' in root folder of your project

File: Task_4/main.go [hunk 1]
```diff
@@ -1,10 +1,8 @@
 package main
 
 import (
-
 	"structures/shapes"
-
-)
+	)
 
 func main() {
```
Comment (rat6666) [inline Task_4/main.go]: remove please extra empty line

File: Task_4/shapes/circle.go [hunk 2]
```diff
@@ -10,13 +10,22 @@ type Circle struct {
 }
 
 func(c Circle) Area() float64{
+	if c.Radius<0{
```
Comment (rat6666) [inline Task_4/shapes/circle.go]: code not formated (missed space)

File: Task_4/shapes/circle.go [hunk 3]
```diff
@@ -10,13 +10,22 @@ type Circle struct {
...
+		return 0
+	}
 	return math.Pi*c.Radius *c.Radius
 }
 
 func(c Circle) Perimeter() float64 {
+	if c.Radius<0{
```
Comment (rat6666) [inline Task_4/shapes/circle.go]: function should return error in case if values invalid

File: Task_4/shapes/circle_test.go [hunk 4]
```diff
@@ -0,0 +1,68 @@
+package shapes
+
+import (
+	"math"
+	"testing"
+)
+
+func TestCircle_Area(t *testing.T) {
+
```
Comment (rat6666) [inline Task_4/shapes/circle_test.go]: extra empty line

File: Task_4/shapes/circle_test.go [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+			}
+			if got := c.Area(); got != tt.want {
+				t.Errorf("Area() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+func TestCircle_Perimeter(t *testing.T) {
+
```
Comment (rat6666) [inline Task_4/shapes/circle_test.go]: extra empty line

File: Task_4/shapes/rectangle.go [hunk 6]
```diff
@@ -8,13 +8,22 @@ type Rectangle struct {
 }
 
 func(r Rectangle) Area() float64{
+	if r.Height<0 || r.Width<0{
```
Comment (rat6666) [inline Task_4/shapes/rectangle.go]: function should return error

File: Task_4/shapes/rectangle_test.go [hunk 7]
```diff
@@ -0,0 +1,70 @@
+package shapes
+
+import "testing"
+
+func TestRectangle_Area(t *testing.T) {
+
```
Comment (rat6666) [inline Task_4/shapes/rectangle_test.go]: emptyy extra line

File: Task_4/shapes/rectangle_test.go [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+			}
+			if got := r.Area(); got != tt.want {
+				t.Errorf("Area() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+func TestRectangle_Perimeter(t *testing.T) {
+
```
Comment (rat6666) [inline Task_4/shapes/rectangle_test.go]: the same here

**Review by rat6666 — CHANGES_REQUESTED** [review]

**Review by rat6666 — APPROVED** [review]

**Pull request merged**